packages/webchannel-wrapper/src/index.d.ts

```diff
@@ -15,23 +15,65 @@
  * limitations under the License.
  */
 
-// TODO(mikelehen): Flesh out proper types (or figure out how to generate via
-// clutz or something).
-export class XhrIo {}
-export const ErrorCode: any;
-export const EventType: any;
+// WARNING: This is not a complete set of types exported by WebChannelWrapper.
+// It is merely meant to support the usage patterns of the Firestore SDK.
+
+export var EventType: {
+  COMPLETE: string;
+};
+
 export namespace WebChannel {
-  export type EventType = any;
-  export const EventType: any;
+  export var EventType: {
+    OPEN: string;
+    CLOSE: string;
+    ERROR: string;
+    MESSAGE: string;
+  };
+}
+
+export var ErrorCode: {
+  NO_ERROR: number;
+  HTTP_ERROR: number;
+  TIMEOUT: number;
+};
+
+export interface Headers {
+  [name: string]: string | number;
+}
+
+export interface WebChannelError {
+  error?: { status: string; message: string };
 }
 
-type StringMap = { [key: string]: string };
+export class XhrIo {
+  send(
+    url: string,
+    method?: string,
+    body?: string,
+    headers?: Headers,
+    timeoutInterval?: number
+  ): void;
+
+  getLastErrorCode(): number;
+
+  getLastError(): string;
+
+  getStatus(): number;
+
+  getResponseText(): string;
+
+  getResponseJson(): WebChannelError | object;
+
+  listenOnce<T>(type: string, cb: (param: T) => void): void;
+}
 
 export interface WebChannelOptions {
-  messageHeaders?: StringMap;
-  initMessageHeaders?: StringMap;
+  messageHeaders?: {};
+  initMessageHeaders?: {};
```

Comment:
Is your intention to require that we set these to an empty object?  Because I don't think that's what this actually does.  This will allow us to pass any object (e.g. `{'foo': 'bar'}` since TypeScript does structural typing (aka duck typing).  

Not sure it's possible to specify a type that indicates that only an empty object is valid.

That said, in general this is feeling somewhat fragile which is a bit worrisome since webchannel is already a bit fragile. It wouldn't be too surprising if we tweak some webchannel setting for performance/reliability reasons and then it gets minified wrong because we don't properly update this d.ts file and therefore it's ignored, and we don't notice because it's not something we can readily test.

If that's the world we're heading towards, I'd like to discuss if there are ways to avoid it...  and if not, then we will perhaps eventually (once your minification stuff lands) want some big warnings in this file or something.

Reply:
To be perfectly honest, that was my intention. I did realize this doesn't work but I thought that it might still convey some of the original intent. 

But you can ignore all this because I came up with a way that actually works. I can disallow all properties with:

`messageHeaders: { [k : string] : never }`

Furthermore, all members that have at least a single property specified do not allow any extra properties as is.